File: 238-Product-of-Array-Except-Self/solution.go
package main

func productExceptSelf(nums []int) []int {
	res := make([]int, len(nums))

	// calculate the product of numbers to the left of nums[i]
	prefix := 1
	for idx, val := range nums {
		res[idx] = prefix
		prefix *= val
	}

	// Although this value is used as a suffix, we assign it to
	// 'prefix' to optimize time and memory.
	prefix = 1

	// calculate the product of numbers to the right of nums[i]
	for i := len(nums) - 1; i >= 0; i-- {
		res[i] *= prefix
		prefix *= nums[i]
	}

	return res
}

File: 238-Product-of-Array-Except-Self/solution.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var productExceptSelf = function (nums) {
  const res = new Array(nums.length);
  res[0] = 1;

  // calculate the product of all the numbers to to left of nums[i]
  for (let i = 1; i < nums.length; i++) {
    res[i] = res[i - 1] * nums[i - 1];
  }

  let suffix = 1;

  // calculate the product of all the numbers to the right of nums[i].
  // This is done by using a 'suffix' to keep track of the product
  //   of all elements to the right of nums[i]
  for (let i = res.length - 1; i >= 0; i--) {
    res[i] *= suffix;
    suffix *= nums[i];
  }

  return res;
};

productExceptSelf([1, 2, 3, 4]);
